firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -112,14 +128,74 @@ public void onFailure(@NonNull @NotNull Exception e) {
     return taskCompletionSource.getTask();
   }
 
+  /** Signs in the App Distribution tester. Presents the tester with a Google sign in UI */
+  @NonNull
+  public Task<Void> signInTester() {
+    if (signInTaskCompletionSource != null && !signInTaskCompletionSource.getTask().isComplete()) {
+      signInCancellationSource.cancel();
```

Comment:
I think the `signInCancellationSource` may be null here.  My suggestion is to pull the SignIn and the associated mutable members into a new class, something like:


```
public class TesterSignIn {
...
...

public TesterSignIn(Context context) {
    this.context = context;
    this.cancelizationSource = ...
}

public Task<Void> signInTester() {
 cancelInProcessSignIn();
 storeNewCancelSource();
 showSigninDialog()
}
}
```

Reply:
signInCancellationSource shouldn't be null whenever signInTaskCompletionSource isn't null either since they are both initialized together after this check, so I think that checking that ```signInTaskCompletionSource != null``` before cancelling protects this from being null. I agree that putting checkForUpdate and signIn in their own classes would be better for modularity and for readability

Reply:
In a cases like this, where there are multiple parameters, but a smaller logical state-space (e.g. signInTask shouldn't be null if isSigningIn = true), that's a signal that state that mutates together is ripe for getting encapsulated together and then mutated only behind the encapsulation.